### PR TITLE
Add a 'to' method for moving to and from device for BlockMask

### PR DESCRIPTION
**Pull request by drisspg**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #132087

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: test/inductor/test_flex_attention.py [hunk 1]
```diff
@@ -1505,6 +1505,17 @@ def causal_mask(b, h, q, kv):
...
+    @supported_platform
+    def test_block_mask_device_change(self):
+        offset = torch.zeros(8, device="cuda")
+
+        def causal_mask(b, h, q, kv):
+            return (q + (offset[b] * 128)) >= kv
+
+        block_mask = create_block_mask(causal_mask, 1, 1, 512, 512)
+        block_mask = block_mask.to("cpu")
+        block_mask = block_mask.to("cuda")
```
Comment (RdoubleA) [inline test/inductor/test_flex_attention.py:1517]: do you need to assert anything here? maybe checking if it switched to CPU and if it switched back to cuda?

**Review by RdoubleA — COMMENTED** [review]
Thank you for adding this at my request :) Are the `Q_BLOCK_SIZE`/`KV_BLOCK_SIZE` intended to be included with the `to()` change?

File: test/inductor/test_flex_attention.py (hunk 1 above)
Comment (drisspg) [reply test/inductor/test_flex_attention.py:1527]: just added

**Comment by drisspg** [comment]
We used to have them unpacked and then we recently packed them into the tuple. It was more annoying to have to add some special logic just for the blocksize when, unpacking -> applying func -> repacking. I imagine Horace might complain but I will wait for him to :)

**Comment by Chillee** [comment]
@drisspg I'd prefer for them to be packed - it makes it slightly easier to create a custom blockmask generator.

**Review by yanboliang — APPROVED** [review]

File: torch/nn/attention/flex_attention.py [hunk 2]
```diff
@@ -253,34 +254,58 @@ def __init__(
...
             self.full_q_num_blocks, self.full_q_indices = _transpose_ordered(
                 full_kv_num_blocks, full_kv_indices
             )
         else:
             self.full_q_num_blocks, self.full_q_indices = None, None
+
         if isinstance(BLOCK_SIZE, int):
             BLOCK_SIZE = (BLOCK_SIZE, BLOCK_SIZE)
         self.BLOCK_SIZE = BLOCK_SIZE
-        if mask_mod is None:
```
Comment (drisspg) [inline torch/nn/attention/flex_attention.py:279]: Todo, this is a code mistake

**Pull request closed without merging**